components/brave_wallet/browser/bitcoin/bitcoin_rpc.cc

```diff
@@ -0,0 +1,241 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/bitcoin/bitcoin_rpc.h"
+
+#include "base/command_line.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom-forward.h"
+#include "brave/components/brave_wallet/common/switches.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace {
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("bitcoin_rpc", R"(
+      semantics {
+        sender: "Bitcoin RPC"
+        description:
+          "This service is used to communicate with Bitcoin nodes "
+          "on behalf of the user interacting with the native Brave wallet."
+        trigger:
+          "Triggered by uses of the native Brave wallet."
+        data:
+          "Bitcoin JSON RPC response bodies."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable this feature on chrome://flags."
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+const GURL MakeGetChainHeightUrl(const GURL& base_url) {
+  GURL::Replacements replacements;
+  const std::string path =
+      base_url.path() + base::JoinString({"blocks", "tip", "height"}, "/");
+  replacements.SetPathStr(path);
+
+  return base_url.ReplaceComponents(replacements);
+}
+
+const GURL MakeAddressHistoryUrl(const GURL& base_url,
+                                 const std::string& address,
+                                 const std::string& last_seen_txid) {
+  GURL::Replacements replacements;
+  std::string path =
+      base_url.path() +
+      base::JoinString({"address", address, "txs", "chain"}, "/");
+  if (!last_seen_txid.empty()) {
+    path += "/" + last_seen_txid;
+  }
+  replacements.SetPathStr(path);
+
+  return base_url.ReplaceComponents(replacements);
+}
+
+const GURL MakePostTransactionUrl(const GURL& base_url) {
+  GURL::Replacements replacements;
+  const std::string path = base_url.path() + "tx";
+  replacements.SetPathStr(path);
+
+  return base_url.ReplaceComponents(replacements);
+}
+
+std::string BitcoinRpcHost() {
+  return base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII(
+      brave_wallet::switches::kBitcoinRpcHost);
+}
+
+GURL BaseRpcUrl(const std::string& network_id) {
+  CHECK(brave_wallet::IsBitcoinNetwork(network_id));
+
+  std::string host = BitcoinRpcHost();
+  std::string path = network_id == brave_wallet::mojom::kBitcoinMainnet
```

Comment:
Can we verify that the network ID follows a network ID pattern somewhere?
Do we already?

Reply:
Not sure I get the question right.
These are internal names so we can differentiate between main/test networks

Reply:
I'm most worried that somehow these arguments may be in `../../../0xformat` and when string appended may generate an API path traversal

Reply:
network id in that line is used to pick fixed strings: either "/api/" or "/testnet/api/" as base api path.
However in this file transaction ids and bitcoin addresses are also used to combine request urls but they are supposed to be strings like `a1075db55d416d3ca199f55b6084e2115b9345e16c5cf302fc80e9d5fbf5d48d` or `tb1qmnn0dpgg3qjmdkh6vg0p743q9jrlt9nmc5as4f`. 
I can add some sanitization to ensure that URLs are combined with only alphanumeric runtime strings.

Reply:
It would be perfect, thanks!